donkeycar/parts/tachometer.py

```diff
@@ -0,0 +1,732 @@
+        self.pin = gpio_pin
+        self.debounce_ns:int = debounce_ns
+        self.debounce_time:int = 0
+        if self.debounce_ns > 0:
+            logger.warn("GpioEncoder debounce_ns will be ignored.")
+        self.lock = threading.Lock()
+
+    def _cb(self):
+        """
+        Callback routine called by GPIO when a tick is detected
+        :pin_number: int the pin number that generated the interrupt.
+        :pin_state: int the state of the pin
+        """
+        #
+        # we avoid blocking by updating an internal counter,
+        # then if we can get a lock, use the internal counter
+        # to update the public counter used by the application.
+        #
+        self._cb_counter += 1
+        if self.lock.acquire(blocking=False):
+            try:
+                self.counter += self._cb_counter * self.direction
+                self._cb_counter = 0
+            finally:
+                self.lock.release()
+            
+    def start_ticks(self):
+        # configure GPIO pin
+        self.pin.start(on_input=lambda: self._cb(), edge=PinEdge.RISING)
+        logger.info(f'GpioEncoder on InputPin "RPI_GPIO.{self.pin.pin_scheme_str}.{self.pin.pin_number}" started.')
+
+    def poll_ticks(self, direction:int):  
+        """
+        read the encoder ticks
+        direction: 1 if forward, -1 if reverse, 0 if stopped.
+        return: updated encoder ticks
+        """
+        with self.lock:
+            self.direction = direction
+
+    def stop_ticks(self):
+        self.pin.stop()
+        logger.info(f'GpioEncoder on InputPin "RPI_GPIO.{self.pin.pin_scheme_str}.{self.pin.pin_number}" stopped.')
+
+    def get_ticks(self, encoder_index:int=0) -> int:
+        """
+        Get last polled encoder ticks
+        encoder_index: zero based index of encoder.
+        return: Most recently polled encoder ticks
+
+        This will return the same value as the
+        most recent call to poll_ticks().  It 
+        will not request new values from the encoder.
+        This will not block.
+        """
+        with self.lock:
+            return self.counter if encoder_index == 0 else 0
+
+
+class MockEncoder(AbstractEncoder):
+    """
+    A mock encoder that turns throttle values into ticks.
+    It generates ENCODER_PPR ticks per second at full throttle.
+    The run() method must be called at the same rate as the
+    tachometer calls the poll() method.
+    """
+    def __init__(self, ticks_per_second: float):
+        self.ticks_per_second = ticks_per_second
+        self.throttle = 0
+        self.ticks = 0
+        self.remainder_ticks = 0
+        self.timestamp = None
+        self.running = False
+
+    def run(self, throttle:float, timestamp: int = None):
+        if timestamp is None:
+            timestamp = time.time()
+
+        #
+        # poll() passed None for throttle and steering,
+        # so we use the last value passed to run()
+        #
+        if throttle is not None:
```

Comment:
The change was made; I'm not sure why the related comment is still open

Reply:
Ok, looks good now.